lightning_uq_box/uq_methods/laplace_model.py

```diff
@@ -28,29 +27,44 @@
 # over training_step in lightning
 
 
-def tune_prior_precision(
-    model: Laplace, tune_precision_lr: float, n_epochs_tune_precision: int
+def tune_prior_precision_and_sigma(
+    model: Laplace,
+    tune_precision_lr: float,
+    n_epochs_tune_precision: int,
+    tune_prior_precision: bool,
+    tune_sigma_noise: bool,
 ):
-    """Tune the prior precision via Empirical Bayes.
+    """Tune the prior precision and sigma noise via Empirical Bayes.
 
     Args:
         model: laplace model
         tune_precision_lr: learning rate for tuning prior precision
         n_epochs_tune_precision: number of epochs to tune prior precision
+        tune_prior_precision: whether to tune prior precision
+        tune_sigma_noise: whether to tune sigma noise
     """
-    log_prior, log_sigma = (
-        torch.ones(1, requires_grad=True),
-        torch.ones(1, requires_grad=True),
-    )
-    hyper_optimizer = torch.optim.Adam([log_prior, log_sigma], lr=tune_precision_lr)
-    bar = trange(n_epochs_tune_precision)
-    # find out why this is so extremely slow?
-    for i in bar:
-        hyper_optimizer.zero_grad()
-        neg_marglik = -model.log_marginal_likelihood(log_prior.exp(), log_sigma.exp())
-        neg_marglik.backward()
-        hyper_optimizer.step()
-        bar.set_postfix(neg_marglik=f"{neg_marglik.detach().cpu().item()}")
+    with torch.inference_mode(False):
+        log_prior, log_sigma = (
+            torch.ones(1, requires_grad=True),
+            torch.ones(1, requires_grad=True),
+        )
+        optim_params = []
+        if model.prior_precision is not None:
+            optim_params.append(log_prior)
+        elif model.sigma_noise is not None:
+            optim_params.append(log_sigma)
```

Comment:
I guess this is the place where `tune_prior_precision` and `tune_sigma_noise` should be used? 
I *think* that `la.prior_precision` and `la.sigma_noise` are tensor-valued properties which are never None. I'm not sure what the reason for the if ... elseif is. This would only ever optimize one of the parameters. If both are never None, then this would always trigger `prior_precision` optimization.